dev-tools/ls-apis/src/system_apis.rs

```diff
@@ -383,16 +401,322 @@ impl SystemApis {
+                    // case (because we will catch that case when we're
+                    // iterating over the dependency API).
+                    if api.versioned_how == VersionedHow::Server
+                        && dependency_api.versioned_how == VersionedHow::Unknown
+                    {
+                        dag_check.propose_client(
+                            dependency_clientpkg,
+                            format!(
+                                "has cyclic dependency on {:?}, which is \
+                                 server-managed",
+                                api.client_package_name,
+                            ),
+                        )
+                    }
+
+                    // If both are Unknown, tell the user to pick one.
+                    if api.versioned_how == VersionedHow::Unknown
+                        && dependency_api.versioned_how == VersionedHow::Unknown
+                    {
+                        dag_check.propose_upick(
+                            &api.client_package_name,
+                            dependency_clientpkg,
+                        );
+                    }
+                }
+            }
+        }
+
+        Ok(dag_check)
+    }
+}
+
+pub struct DagCheck<'a> {
+    proposed_server_managed: BTreeMap<&'a ClientPackageName, Vec<String>>,
+    proposed_client_managed: BTreeMap<&'a ClientPackageName, Vec<String>>,
+    proposed_upick:
+        BTreeMap<&'a ClientPackageName, BTreeSet<&'a ClientPackageName>>,
+}
+
+impl<'a> DagCheck<'a> {
+    fn new() -> DagCheck<'a> {
+        DagCheck {
+            proposed_server_managed: BTreeMap::new(),
+            proposed_client_managed: BTreeMap::new(),
+            proposed_upick: BTreeMap::new(),
+        }
+    }
+
+    fn propose_client(
+        &mut self,
+        client_pkgname: &'a ClientPackageName,
+        reason: String,
+    ) {
+        self.proposed_client_managed
+            .entry(client_pkgname)
+            .or_insert_with(Vec::new)
+            .push(reason);
+    }
+
+    fn propose_server(
+        &mut self,
+        client_pkgname: &'a ClientPackageName,
+        reason: String,
+    ) {
+        self.proposed_server_managed
+            .entry(client_pkgname)
+            .or_insert_with(Vec::new)
+            .push(reason);
+    }
+
+    fn propose_upick(
+        &mut self,
+        client_pkgname1: &'a ClientPackageName,
+        client_pkgname2: &'a ClientPackageName,
+    ) {
+        // Avoid duplicates.  (It's easier for us to do this here than for the
+        // caller to avoid this case.)
+        if let Some(other_pkg_upicks) = self.proposed_upick.get(client_pkgname2)
+        {
+            if other_pkg_upicks.contains(client_pkgname1) {
+                return;
+            }
+        }
+
+        self.proposed_upick
+            .entry(client_pkgname1)
+            .or_insert_with(BTreeSet::new)
+            .insert(client_pkgname2);
```

Comment:
Hmm I might be misreading but it looks like the first `get` happens with `pkgname2` but the second `entry` happens with `pkgname1`? Is this intentional -- if so it might be worth explaining what's going on here.

Unless the goal here is to represent unordered pairs of packages, in which case maybe this can be a `BTreeSet<(&'a ClientPackageName, &'a ClientPackageName)>` where you check in both directions? Or maybe even a [`BiMap`](https://docs.rs/bimap/latest/bimap/) which is kind of geared towards this.

Reply:
> Unless the goal here is to represent unordered pairs of packages, in which case maybe this can be a BTreeSet<(&'a ClientPackageName, &'a ClientPackageName)> where you check in both directions? Or maybe even a [BiMap](https://docs.rs/bimap/latest/bimap/) which is kind of geared towards this.

Yeah, a "upick" here is a situation where the tool has found that p1 and p2 depend on each other, so you (the person running the tool) must pick one of them to be client-managed.  The mapping here is just to prevent it from reporting this message twice: once when looking at p1 and once when looking at p2.  I don't think `BiMap` works because this isn't bijective: I think you could have a "upick" between p1/p2 as well as one between p2/p3.  What's here should be equivalent to `BTreeSet<(&'a ClientPackageName, &'a ClientPackageName)>` and I'm not sure why I went the way I did but neither seems obviously better to me.

I'll at least document this better.